Reals intersect with Range version of Integers fails

```julia
In [1]: Reals.intersect(Range(-oo, oo, 1))                                                                                                                    
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
~/current/sympy/upstream.sympy.git/sympy/solvers/diophantine.py in classify_diop(eq, _dict)
    479         var = list(eq.free_symbols)
--> 480         assert var
    481     except (AttributeError, AssertionError):

AssertionError: 

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-1-42f8934c336b> in <module>
----> 1 Reals.intersect(Range(-oo, oo, 1))

~/current/sympy/upstream.sympy.git/sympy/sets/sets.py in intersect(self, other)
    138 
    139         """
--> 140         return Intersection(self, other)
    141 
    142     def intersection(self, other):

~/current/sympy/upstream.sympy.git/sympy/sets/sets.py in __new__(cls, *args, **kwargs)
   1309         if evaluate:
   1310             args = list(cls._new_args_filter(args))
-> 1311             return simplify_intersection(args)
   1312 
   1313         args = list(ordered(args, Set._infimum_key))

~/current/sympy/upstream.sympy.git/sympy/sets/sets.py in simplify_intersection(args)
   2210             new_args = False
   2211             for t in args - set((s,)):
-> 2212                 new_set = intersection_sets(s, t)
   2213                 # This returns None if s does not know how to intersect
   2214                 # with t. Returns the newly intersected set otherwise

~/current/sympy/upstream.sympy.git/sympy/multipledispatch/dispatcher.py in __call__(self, *args, **kwargs)
    196             self._cache[types] = func
    197         try:
--> 198             return func(*args, **kwargs)
    199 
    200         except MDNotImplementedError:

~/current/sympy/upstream.sympy.git/sympy/sets/handlers/intersection.py in intersection_sets(a, b)
     97     if end not in b:
     98         end -= 1
---> 99     return intersection_sets(a, Range(start, end + 1))
    100 
    101 @dispatch(Range, Naturals)

~/current/sympy/upstream.sympy.git/sympy/multipledispatch/dispatcher.py in __call__(self, *args, **kwargs)
    196             self._cache[types] = func
    197         try:
--> 198             return func(*args, **kwargs)
    199 
    200         except MDNotImplementedError:

~/current/sympy/upstream.sympy.git/sympy/sets/handlers/intersection.py in intersection_sets(a, b)
    134     # have integer solutions so we use the diophantine
    135     # solver
--> 136     va, vb = diop_linear(eq(r1, Dummy('a')) - eq(r2, Dummy('b')))
    137 
    138     # check for no solution

~/current/sympy/upstream.sympy.git/sympy/solvers/diophantine.py in diop_linear(eq, param)
    632     diop_general_sum_of_squares()
    633     """
--> 634     var, coeff, diop_type = classify_diop(eq, _dict=False)
    635 
    636     if diop_type == "linear":

~/current/sympy/upstream.sympy.git/sympy/solvers/diophantine.py in classify_diop(eq, _dict)
    480         assert var
    481     except (AttributeError, AssertionError):
--> 482         raise ValueError('equation should have 1 or more free symbols')
    483     var.sort(key=default_sort_key)
    484     eq = eq.expand(force=True)

ValueError: equation should have 1 or more free symbols
```

--- a/sympy/calculus/tests/test_singularities.py
+++ b/sympy/calculus/tests/test_singularities.py
@@ -46,6 +46,8 @@ def test_is_increasing():
     assert is_increasing(-x**2*a, Interval(1, oo), x)
     assert is_increasing(1)
 
+    assert is_increasing(4*x**3 - 6*x**2 - 72*x + 30, Interval(-2, 3)) is False
+
 
 def test_is_strictly_increasing():
     """Test whether is_strictly_increasing returns correct value."""
@@ -58,6 +60,7 @@ def test_is_strictly_increasing():
     assert not is_strictly_increasing(-x**2, Interval(0, oo))
     assert not is_strictly_decreasing(1)
 
+    assert is_strictly_increasing(4*x**3 - 6*x**2 - 72*x + 30, Interval.open(-2, 3)) is False
 
 def test_is_decreasing():
     """Test whether is_decreasing returns correct value."""

--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -40,6 +40,7 @@ class Rationals(with_metaclass(Singleton, Set)):
     _inf = S.NegativeInfinity
     _sup = S.Infinity
     is_empty = False
+    is_finiteset = False
 
     def _contains(self, other):
         if not isinstance(other, Expr):
@@ -105,7 +106,7 @@ class Naturals(with_metaclass(Singleton, Set)):
     _inf = S.One
     _sup = S.Infinity
     is_empty = False
-    is_finite = False
+    is_finiteset = False
 
     def _contains(self, other):
         if not isinstance(other, Expr):
@@ -141,7 +142,6 @@ class Naturals0(Naturals):
     Integers : also includes the negative integers
     """
     _inf = S.Zero
-    is_empty = False
 
     def _contains(self, other):
         if not isinstance(other, Expr):
@@ -189,6 +189,7 @@ class Integers(with_metaclass(Singleton, Set)):
 
     is_iterable = True
     is_empty = False
+    is_finiteset = False
 
     def _contains(self, other):
         if not isinstance(other, Expr):

--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -48,6 +48,10 @@ class Rationals(with_metaclass(Singleton, Set)):
             return other.is_Rational
         return other.is_rational
 
+    def _eval_is_subset(self, other):
+        if other.is_subset(S.Integers):
+            return False
+
     def __iter__(self):
         from sympy.core.numbers import igcd, Rational
         yield S.Zero
@@ -146,6 +150,10 @@ class Naturals0(Naturals):
         elif other.is_integer is False or other.is_nonnegative is False:
             return S.false
 
+    def _eval_is_subset(self, other):
+        if other.is_subset(S.Naturals):
+            return False
+
 
 class Integers(with_metaclass(Singleton, Set)):
     """
@@ -210,6 +218,10 @@ class Integers(with_metaclass(Singleton, Set)):
         from sympy.functions.elementary.integers import floor
         return And(Eq(floor(x), x), -oo < x, x < oo)
 
+    def _eval_is_subset(self, other):
+        if other.is_subset(S.Naturals0):
+            return False
+
 
 class Reals(with_metaclass(Singleton, Interval)):
     """
@@ -248,6 +260,10 @@ class Reals(with_metaclass(Singleton, Interval)):
 
     def __hash__(self):
         return hash(Interval(S.NegativeInfinity, S.Infinity))
+
+    def _eval_is_subset(self, other):
+        if other.is_subset(S.Rationals):
+            return False
 
 
 class ImageSet(Set):
@@ -1375,6 +1391,8 @@ class Complexes(with_metaclass(Singleton, CartesianComplexRegion)):
 
     """
 
+    is_empty = False
+
     # Override property from superclass since Complexes has no args
     sets = ProductSet(S.Reals, S.Reals)
 
@@ -1386,3 +1404,7 @@ class Complexes(with_metaclass(Singleton, CartesianComplexRegion)):
 
     def __repr__(self):
         return "S.Complexes"
+
+    def _eval_is_subset(self, other):
+        if other.is_subset(S.Reals):
+            return False

--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -40,7 +40,7 @@ class Rationals(with_metaclass(Singleton, Set)):
     _inf = S.NegativeInfinity
     _sup = S.Infinity
     is_empty = False
-    is_finiteset = False
+    is_finite_set = False
 
     def _contains(self, other):
         if not isinstance(other, Expr):
@@ -102,7 +102,7 @@ class Naturals(with_metaclass(Singleton, Set)):
     _inf = S.One
     _sup = S.Infinity
     is_empty = False
-    is_finiteset = False
+    is_finite_set = False
 
     def _contains(self, other):
         if not isinstance(other, Expr):
@@ -193,7 +193,7 @@ class Integers(with_metaclass(Singleton, Set)):
 
     is_iterable = True
     is_empty = False
-    is_finiteset = False
+    is_finite_set = False
 
     def _contains(self, other):
         if not isinstance(other, Expr):
@@ -1396,7 +1396,7 @@ class Complexes(with_metaclass(Singleton, CartesianComplexRegion)):
     """
 
     is_empty = False
-    is_finiteset = False
+    is_finite_set = False
 
     # Override property from superclass since Complexes has no args
     sets = ProductSet(S.Reals, S.Reals)

--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -49,10 +49,6 @@ class Rationals(with_metaclass(Singleton, Set)):
             return other.is_Rational
         return other.is_rational
 
-    def _eval_is_subset(self, other):
-        if other.is_subset(S.Integers):
-            return False
-
     def __iter__(self):
         from sympy.core.numbers import igcd, Rational
         yield S.Zero
@@ -229,8 +225,10 @@ class Integers(with_metaclass(Singleton, Set)):
         return And(Eq(floor(x), x), -oo < x, x < oo)
 
     def _eval_is_subset(self, other):
-        if other.is_subset(S.Naturals0):
-            return False
+        return Range(-oo, oo).is_subset(other)
+
+    def _eval_is_superset(self, other):
+        return Range(-oo, oo).is_superset(other)
 
 
 class Reals(with_metaclass(Singleton, Interval)):
@@ -270,10 +268,6 @@ class Reals(with_metaclass(Singleton, Interval)):
 
     def __hash__(self):
         return hash(Interval(S.NegativeInfinity, S.Infinity))
-
-    def _eval_is_subset(self, other):
-        if other.is_subset(S.Rationals):
-            return False
 
 
 class ImageSet(Set):
@@ -1402,6 +1396,7 @@ class Complexes(with_metaclass(Singleton, CartesianComplexRegion)):
     """
 
     is_empty = False
+    is_finiteset = False
 
     # Override property from superclass since Complexes has no args
     sets = ProductSet(S.Reals, S.Reals)
@@ -1414,7 +1409,3 @@ class Complexes(with_metaclass(Singleton, CartesianComplexRegion)):
 
     def __repr__(self):
         return "S.Complexes"
-
-    def _eval_is_subset(self, other):
-        if other.is_subset(S.Reals):
-            return False

--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -105,6 +105,7 @@ class Naturals(with_metaclass(Singleton, Set)):
     _inf = S.One
     _sup = S.Infinity
     is_empty = False
+    is_finite = False
 
     def _contains(self, other):
         if not isinstance(other, Expr):

--- a/sympy/sets/fancysets.py
+++ b/sympy/sets/fancysets.py
@@ -116,6 +116,12 @@ class Naturals(with_metaclass(Singleton, Set)):
         elif other.is_integer is False or other.is_positive is False:
             return False
 
+    def _eval_is_subset(self, other):
+        return Range(1, oo).is_subset(other)
+
+    def _eval_is_superset(self, other):
+        return Range(1, oo).is_superset(other)
+
     def __iter__(self):
         i = self._inf
         while True:
@@ -152,8 +158,10 @@ class Naturals0(Naturals):
             return S.false
 
     def _eval_is_subset(self, other):
-        if other.is_subset(S.Naturals):
-            return False
+        return Range(oo).is_subset(other)
+
+    def _eval_is_superset(self, other):
+        return Range(oo).is_superset(other)
 
 
 class Integers(with_metaclass(Singleton, Set)):

--- a/sympy/sets/handlers/intersection.py
+++ b/sympy/sets/handlers/intersection.py
@@ -126,6 +126,13 @@ def intersection_sets(a, b):
     if r2.start.is_infinite:
         r2 = r2.reversed
 
+    # If both ends are infinite then it means that one Range is just the set
+    # of all integers (the step must be 1).
+    if r1.start.is_infinite:
+        return b
+    if r2.start.is_infinite:
+        return a
+
     # this equation represents the values of the Range;
     # it's a linear equation
     eq = lambda r, i: r.start + i*r.step

--- a/sympy/sets/handlers/issubset.py
+++ b/sympy/sets/handlers/issubset.py
@@ -2,7 +2,7 @@ from sympy import S
 from sympy.core.logic import fuzzy_and, fuzzy_bool, fuzzy_not, fuzzy_or
 from sympy.core.relational import Eq
 from sympy.sets.sets import FiniteSet, Interval, Set, Union
-from sympy.sets.fancysets import Reals, Range
+from sympy.sets.fancysets import Complexes, Naturals, Reals, Range, Rationals
 from sympy.multipledispatch import dispatch
 
 
@@ -52,3 +52,50 @@ def is_subset_sets(a, b):
     if a.step == b.step == 1:
         return fuzzy_and([fuzzy_bool(a.start >= b.start),
                           fuzzy_bool(a.stop <= b.stop)])
+
+@dispatch(Range, Interval)
+def is_subset_sets(a_range, b_interval):
+    if a_range.step.is_positive:
+        if b_interval.left_open and a_range.inf.is_finite:
+            cond_left = a_range.inf > b_interval.left
+        else:
+            cond_left = a_range.inf >= b_interval.left
+        if b_interval.right_open and a_range.sup.is_finite:
+            cond_right = a_range.sup < b_interval.right
+        else:
+            cond_right = a_range.sup <= b_interval.right
+        return fuzzy_and([cond_left, cond_right])
+
+@dispatch(Interval, Range)
+def is_subset_sets(a_interval, b_range):
+    if a_interval.measure.is_extended_nonzero:
+        return False
+
+@dispatch(Interval, Rationals)
+def is_subset_sets(a_interval, b_rationals):
+    if a_interval.measure.is_extended_nonzero:
+        return False
+
+@dispatch(Range, Complexes)
+def is_subset_sets(a, b):
+    return True
+
+@dispatch(Complexes, Interval)
+def is_subset_sets(a, b):
+    return False
+
+@dispatch(Complexes, Range)
+def is_subset_sets(a, b):
+    return False
+
+@dispatch(Complexes, Rationals)
+def is_subset_sets(a, b):
+    return False
+
+@dispatch(Rationals, Reals)
+def is_subset_sets(a, b):
+    return True
+
+@dispatch(Rationals, Range)
+def is_subset_sets(a, b):
+    return False

--- a/sympy/sets/handlers/issubset.py
+++ b/sympy/sets/handlers/issubset.py
@@ -2,7 +2,7 @@ from sympy import S
 from sympy.core.logic import fuzzy_and, fuzzy_bool, fuzzy_not, fuzzy_or
 from sympy.core.relational import Eq
 from sympy.sets.sets import FiniteSet, Interval, Set, Union
-from sympy.sets.fancysets import Reals
+from sympy.sets.fancysets import Reals, Range
 from sympy.multipledispatch import dispatch
 
 
@@ -10,9 +10,7 @@ _inf_sets = [S.Naturals, S.Naturals0, S.Integers, S.Rationals, S.Reals, S.Comple
 
 @dispatch(Set, Set)
 def is_subset_sets(a, b):
-    if a in _inf_sets:
-        if b in _inf_sets:
-            return _inf_sets.index(a) <= _inf_sets.index(b)
+    return None
 
 @dispatch(Interval, Interval)
 def is_subset_sets(a, b):
@@ -48,3 +46,9 @@ def is_subset_sets(a_interval, b_u):
                     ])
             if all(no_overlap(s, a_interval) for s in intervals):
                 return False
+
+@dispatch(Range, Range)
+def is_subset_sets(a, b):
+    if a.step == b.step == 1:
+        return fuzzy_and([fuzzy_bool(a.start >= b.start),
+                          fuzzy_bool(a.stop <= b.stop)])

--- a/sympy/sets/handlers/issubset.py
+++ b/sympy/sets/handlers/issubset.py
@@ -1,0 +1,60 @@
+from sympy import S
+from sympy.core.logic import fuzzy_and, fuzzy_bool, fuzzy_not, fuzzy_or
+from sympy.core.relational import Eq
+from sympy.sets.sets import FiniteSet, Interval, Set, Union
+from sympy.sets.fancysets import Reals
+from sympy.multipledispatch import dispatch
+
+
+_inf_sets = [S.Naturals, S.Naturals0, S.Integers, S.Rationals, S.Reals, S.Complexes]
+
+@dispatch(Set, Set)
+def is_subset_sets(a, b):
+    if a in _inf_sets:
+        if b in _inf_sets:
+            return _inf_sets.index(a) <= _inf_sets.index(b)
+        if isinstance(b, FiniteSet):
+            return False
+
+@dispatch(Interval, Interval)
+def is_subset_sets(a, b):
+    # This is correct but can be made more comprehensive...
+    if fuzzy_bool(a.start < b.start):
+        return False
+    if fuzzy_bool(a.end > b.end):
+        return False
+    if (b.left_open and not a.left_open and fuzzy_bool(Eq(a.start, b.start))):
+        return False
+    if (b.right_open and not a.right_open and fuzzy_bool(Eq(a.end, b.end))):
+        return False
+
+@dispatch(Interval, FiniteSet)
+def is_subset_sets(a_interval, b_fs):
+    # An Interval can only be a subset of a finite set if it is finite
+    # which can only happen if it has zero measure.
+    if fuzzy_not(a_interval.measure.is_zero):
+        return False
+
+@dispatch(Interval, Union)
+def is_subset_sets(a_interval, b_u):
+    if all(isinstance(s, (Interval, FiniteSet)) for s in b_u.args):
+        intervals = [s for s in b_u.args if isinstance(s, Interval)]
+        if all(fuzzy_bool(a_interval.start < s.start) for s in intervals):
+            return False
+        if all(fuzzy_bool(a_interval.end > s.end) for s in intervals):
+            return False
+        if a_interval.measure.is_nonzero:
+            no_overlap = lambda s1, s2: fuzzy_or([
+                    fuzzy_bool(s1.end <= s2.start),
+                    fuzzy_bool(s1.start >= s2.end),
+                    ])
+            if all(no_overlap(s, a_interval) for s in intervals):
+                return False
+
+@dispatch(FiniteSet, Set)
+def is_subset_sets(a_fs, b):
+    return fuzzy_and(b._contains(e) for e in a_fs.args)
+
+@dispatch(Union, Set)
+def is_subset_sets(a_u, b):
+    return fuzzy_and(s.is_subset(b) for s in a_u.args)

--- a/sympy/sets/handlers/issubset.py
+++ b/sympy/sets/handlers/issubset.py
@@ -13,8 +13,6 @@ def is_subset_sets(a, b):
     if a in _inf_sets:
         if b in _inf_sets:
             return _inf_sets.index(a) <= _inf_sets.index(b)
-        if isinstance(b, FiniteSet):
-            return False
 
 @dispatch(Interval, Interval)
 def is_subset_sets(a, b):
@@ -50,11 +48,3 @@ def is_subset_sets(a_interval, b_u):
                     ])
             if all(no_overlap(s, a_interval) for s in intervals):
                 return False
-
-@dispatch(FiniteSet, Set)
-def is_subset_sets(a_fs, b):
-    return fuzzy_and(b._contains(e) for e in a_fs.args)
-
-@dispatch(Union, Set)
-def is_subset_sets(a_u, b):
-    return fuzzy_and(s.is_subset(b) for s in a_u.args)

--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -368,6 +368,9 @@ class Set(Basic):
         ret = self._eval_is_subset(other)
         if ret is not None:
             return ret
+        ret = other._eval_is_superset(self)
+        if ret is not None:
+            return ret
 
         # Use pairwise rules from multiple dispatch
         from sympy.sets.handlers.issubset import is_subset_sets
@@ -383,6 +386,10 @@ class Set(Basic):
             return True
 
     def _eval_is_subset(self, other):
+        '''Returns a fuzzy bool for whether self is a subset of other.'''
+        return None
+
+    def _eval_is_superset(self, other):
         '''Returns a fuzzy bool for whether self is a subset of other.'''
         return None
 

--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -366,7 +366,8 @@ class Set(Basic):
         if not isinstance(other, Set):
             raise ValueError("Unknown argument '%s'" % other)
 
-        if self == other:
+        # Handle the trivial cases
+        if self == other or self.is_empty:
             return True
 
         # Call the subclass method
@@ -379,6 +380,9 @@ class Set(Basic):
             return ret
 
         # Fall back on computing the intersection
+        # XXX: We shouldn't do this. A query like this should be handled
+        # without evaluating new Set objects. It should be the other way round
+        # so that the intersect method uses is_subset for evaluation.
         if self.intersect(other) == self:
             return True
 
@@ -1054,6 +1058,9 @@ class Interval(Set, EvalfMixin):
         return self.as_relational(d).subs(d, other)
 
     def _eval_is_subset(self, other):
+        # Note that self.is_empty was already checked in Set.is_subset and can
+        # be assumed False now.
+
         if isinstance(other, Interval):
             # This is correct but can be made more comprehensive...
             if fuzzy_bool(self.start < other.start):
@@ -1063,9 +1070,18 @@ class Interval(Set, EvalfMixin):
 
         elif isinstance(other, FiniteSet):
             # An Interval can only be a subset of a finite set if it is finite
-            # which can only happen if the start and end are equal.
+            # which can only happen if the start and end are equal (or if it
+            # is empty but that has already been checked).
             if fuzzy_bool(self.start < self.end):
                 return False
+
+        elif isinstance(other, Union):
+            if all(isinstance(s, (Interval, FiniteSet)) for s in other.args):
+                intervals = [s for s in other.args if isinstance(s, Interval)]
+                if all(fuzzy_bool(self.start < s.start) for s in intervals):
+                    return False
+                if all(fuzzy_bool(self.end > s.end) for s in intervals):
+                    return False
 
     def as_relational(self, x):
         """Rewrite an interval in terms of inequalities and logic operators."""

--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -60,7 +60,7 @@ class Set(Basic):
     is_ComplexRegion = False
 
     is_empty = None
-    is_finiteset = None
+    is_finite_set = None
 
     @property
     @deprecated(useinstead="is S.EmptySet or is_empty",
@@ -361,7 +361,7 @@ class Set(Basic):
             return True
         elif fuzzy_not(is_empty) and other.is_empty:
             return False
-        if self.is_finiteset is False and other.is_finiteset:
+        if self.is_finite_set is False and other.is_finite_set:
             return False
 
         # Dispatch on subclass rules
@@ -822,8 +822,8 @@ class ProductSet(Set):
         return fuzzy_or(s.is_empty for s in self.sets)
 
     @property
-    def is_finiteset(self):
-        all_finite = fuzzy_and(s.is_finiteset for s in self.sets)
+    def is_finite_set(self):
+        all_finite = fuzzy_and(s.is_finite_set for s in self.sets)
         return fuzzy_or([self.is_empty, all_finite])
 
     @property
@@ -1023,7 +1023,7 @@ class Interval(Set, EvalfMixin):
         return fuzzy_bool(cond)
 
     @property
-    def is_finiteset(self):
+    def is_finite_set(self):
         return self.measure.is_zero
 
     def _complement(self, other):
@@ -1201,8 +1201,8 @@ class Union(Set, LatticeOp, EvalfMixin):
         return fuzzy_and(set.is_empty for set in self.args)
 
     @property
-    def is_finiteset(self):
-        return fuzzy_and(set.is_finiteset for set in self.args)
+    def is_finite_set(self):
+        return fuzzy_and(set.is_finite_set for set in self.args)
 
     @property
     def _measure(self):
@@ -1356,8 +1356,8 @@ class Intersection(Set, LatticeOp):
         return any(arg.is_iterable for arg in self.args)
 
     @property
-    def is_finiteset(self):
-        if fuzzy_or(arg.is_finiteset for arg in self.args):
+    def is_finite_set(self):
+        if fuzzy_or(arg.is_finite_set for arg in self.args):
             return True
 
     @property
@@ -1586,12 +1586,12 @@ class Complement(Set, EvalfMixin):
             return True
 
     @property
-    def is_finiteset(self):
+    def is_finite_set(self):
         A, B = self.args
-        a_finite = A.is_finiteset
+        a_finite = A.is_finite_set
         if a_finite is True:
             return True
-        elif a_finite is False and B.is_finiteset:
+        elif a_finite is False and B.is_finite_set:
             return False
 
     def __iter__(self):
@@ -1629,7 +1629,7 @@ class EmptySet(with_metaclass(Singleton, Set)):
     .. [1] https://en.wikipedia.org/wiki/Empty_set
     """
     is_empty = True
-    is_finiteset = True
+    is_finite_set = True
     is_FiniteSet = True
 
     @property
@@ -1696,7 +1696,7 @@ class UniversalSet(with_metaclass(Singleton, Set)):
 
     is_UniversalSet = True
     is_empty = False
-    is_finiteset = False
+    is_finite_set = False
 
     def _complement(self, other):
         return S.EmptySet
@@ -1749,7 +1749,7 @@ class FiniteSet(Set, EvalfMixin):
     is_FiniteSet = True
     is_iterable = True
     is_empty = False
-    is_finiteset = True
+    is_finite_set = True
 
     def __new__(cls, *args, **kwargs):
         evaluate = kwargs.get('evaluate', global_evaluate[0])

--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -1067,6 +1067,12 @@ class Interval(Set, EvalfMixin):
                 return False
             if fuzzy_bool(self.end > other.end):
                 return False
+            if (other.left_open and not self.left_open
+                    and fuzzy_bool(Eq(self.start, other.start))):
+                return False
+            if (other.right_open and not self.right_open
+                    and fuzzy_bool(Eq(self.end, other.end))):
+                return False
 
         elif isinstance(other, FiniteSet):
             # An Interval can only be a subset of a finite set if it is finite
@@ -1082,6 +1088,14 @@ class Interval(Set, EvalfMixin):
                     return False
                 if all(fuzzy_bool(self.end > s.end) for s in intervals):
                     return False
+                if self.measure.is_nonzero:
+                    no_overlap = lambda s1, s2: fuzzy_or([
+                            fuzzy_bool(s1.end <= s2.start),
+                            fuzzy_bool(s1.start >= s2.end),
+                            ])
+                    if all(no_overlap(s, self) for s in intervals):
+                        return False
+
 
     def as_relational(self, x):
         """Rewrite an interval in terms of inequalities and logic operators."""

--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -1356,6 +1356,11 @@ class Intersection(Set, LatticeOp):
         return any(arg.is_iterable for arg in self.args)
 
     @property
+    def is_finiteset(self):
+        if fuzzy_or(arg.is_finiteset for arg in self.args):
+            return True
+
+    @property
     def _inf(self):
         raise NotImplementedError()
 
@@ -1579,6 +1584,15 @@ class Complement(Set, EvalfMixin):
     def is_iterable(self):
         if self.args[0].is_iterable:
             return True
+
+    @property
+    def is_finiteset(self):
+        A, B = self.args
+        a_finite = A.is_finiteset
+        if a_finite is True:
+            return True
+        elif a_finite is False and B.is_finiteset:
+            return False
 
     def __iter__(self):
         A, B = self.args

--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -363,25 +363,38 @@ class Set(Basic):
         False
 
         """
-        if isinstance(other, Set):
-            dispatch = getattr(self, '_eval_is_subset', None)
-            if dispatch is not None:
-                ret = dispatch(other)
-                if ret is not None:
-                    return ret
-
-            s_o = self.intersect(other)
-            if s_o == self:
-                return True
-            # This assumes that an unevaluated Intersection will always come
-            # back as an Intersection...
-            elif isinstance(s_o, Intersection):
-                return None
-            else:
-                return False
-        else:
+        if not isinstance(other, Set):
             raise ValueError("Unknown argument '%s'" % other)
 
+        if self == other:
+            return True
+
+        # Call the subclass method
+        ret = self._eval_is_subset(other)
+        if ret is not None:
+            return ret
+
+        ret = other._eval_is_superset(self)
+        if ret is not None:
+            return ret
+
+        # Fall back on computing the intersection
+        if self.intersect(other) == self:
+            return True
+
+    def _eval_is_subset(self, other):
+        """
+        Fuzzy bool indicating whether self is a subset of other
+        """
+        return None
+
+    def _eval_is_superset(self, other):
+        """
+        Fuzzy bool indicating whether self is a superset of other
+        """
+        return None
+
+    # This should be deprecated:
     def issubset(self, other):
         """
         Alias for :meth:`is_subset()`
@@ -426,6 +439,7 @@ class Set(Basic):
         else:
             raise ValueError("Unknown argument '%s'" % other)
 
+    # This should be deprecated:
     def issuperset(self, other):
         """
         Alias for :meth:`is_superset()`
@@ -1039,6 +1053,20 @@ class Interval(Set, EvalfMixin):
         d = Dummy()
         return self.as_relational(d).subs(d, other)
 
+    def _eval_is_subset(self, other):
+        if isinstance(other, Interval):
+            # This is correct but can be made more comprehensive...
+            if fuzzy_bool(self.start < other.start):
+                return False
+            if fuzzy_bool(self.end > other.end):
+                return False
+
+        elif isinstance(other, FiniteSet):
+            # An Interval can only be a subset of a finite set if it is finite
+            # which can only happen if the start and end are equal.
+            if fuzzy_bool(self.start < self.end):
+                return False
+
     def as_relational(self, x):
         """Rewrite an interval in terms of inequalities and logic operators."""
         x = sympify(x)
@@ -1235,6 +1263,13 @@ class Union(Set, LatticeOp, EvalfMixin):
 
     def _contains(self, other):
         return Or(*[s.contains(other) for s in self.args])
+
+    def _eval_is_subset(self, other):
+        return fuzzy_and(s.is_subset(other) for s in self.args)
+
+    def _eval_is_superset(self, other):
+        if fuzzy_or(s.is_superset(other) for s in self.args):
+            return True
 
     def as_relational(self, symbol):
         """Rewrite a Union in terms of equalities and logic operators. """
@@ -1601,6 +1636,12 @@ class EmptySet(with_metaclass(Singleton, Set)):
     def _contains(self, other):
         return false
 
+    def _eval_is_subset(self, other):
+        return True
+
+    def _eval_is_superset(self, other):
+        return other.is_empty
+
     def as_relational(self, symbol):
         return false
 
@@ -1800,6 +1841,14 @@ class FiniteSet(Set, EvalfMixin):
         # evaluate=True is needed to override evaluate=False context;
         # we need Eq to do the evaluation
         return fuzzy_or(fuzzy_bool(Eq(e, other, evaluate=True)) for e in self.args)
+
+    def _eval_is_subset(self, other):
+        return fuzzy_and(other._contains(e) for e in self.args)
+
+    def _eval_is_superset(self, other):
+        # A check for infiniteness would better...
+        if other in [S.Naturals, S.Naturals0, S.Integers, S.Rationals, S.Reals, S.Complexes]:
+            return False
 
     @property
     def _boundary(self):

--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -59,21 +59,8 @@ class Set(Basic):
     is_Complement = None
     is_ComplexRegion = False
 
-    @property
-    def is_empty(self):
-        """
-        Property method to check whether a set is empty.
-        Returns ``True``, ``False`` or ``None`` (if unknown).
-
-        Examples
-        ========
-
-        >>> from sympy import Interval, var
-        >>> x = var('x', real=True)
-        >>> Interval(x, x + 1).is_empty
-        False
-        """
-        return None
+    is_empty = None
+    is_finite = None
 
     @property
     @deprecated(useinstead="is S.EmptySet or is_empty",
@@ -374,6 +361,13 @@ class Set(Basic):
             return True
         elif fuzzy_not(is_empty) and other.is_empty:
             return False
+        if self.is_finite is False and other.is_finite:
+            return False
+
+        # Dispatch on subclass rules
+        ret = self._eval_is_subset(other)
+        if ret is not None:
+            return ret
 
         # Use pairwise rules from multiple dispatch
         from sympy.sets.handlers.issubset import is_subset_sets
@@ -387,6 +381,10 @@ class Set(Basic):
         # so that the intersect method uses is_subset for evaluation.
         if self.intersect(other) == self:
             return True
+
+    def _eval_is_subset(self, other):
+        '''Returns a fuzzy bool for whether self is a subset of other.'''
+        return None
 
     # This should be deprecated:
     def issubset(self, other):
@@ -817,6 +815,11 @@ class ProductSet(Set):
         return fuzzy_or(s.is_empty for s in self.sets)
 
     @property
+    def is_finite(self):
+        all_finite = fuzzy_and(s.is_finite for s in self.sets)
+        return fuzzy_or(self.is_empty, all_finite)
+
+    @property
     def _measure(self):
         measure = 1
         for s in self.sets:
@@ -1243,6 +1246,9 @@ class Union(Set, LatticeOp, EvalfMixin):
 
     def _contains(self, other):
         return Or(*[s.contains(other) for s in self.args])
+
+    def is_subset(self, other):
+        return fuzzy_and(s.is_subset(other) for s in self.args)
 
     def as_relational(self, symbol):
         """Rewrite a Union in terms of equalities and logic operators. """
@@ -1712,6 +1718,7 @@ class FiniteSet(Set, EvalfMixin):
     is_FiniteSet = True
     is_iterable = True
     is_empty = False
+    is_finite = True
 
     def __new__(cls, *args, **kwargs):
         evaluate = kwargs.get('evaluate', global_evaluate[0])
@@ -1808,6 +1815,9 @@ class FiniteSet(Set, EvalfMixin):
         # evaluate=True is needed to override evaluate=False context;
         # we need Eq to do the evaluation
         return fuzzy_or(fuzzy_bool(Eq(e, other, evaluate=True)) for e in self.args)
+
+    def _eval_is_subset(self, other):
+        return fuzzy_and(other._contains(e) for e in self.args)
 
     @property
     def _boundary(self):

--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -60,7 +60,7 @@ class Set(Basic):
     is_ComplexRegion = False
 
     is_empty = None
-    is_finite = None
+    is_finiteset = None
 
     @property
     @deprecated(useinstead="is S.EmptySet or is_empty",
@@ -361,7 +361,7 @@ class Set(Basic):
             return True
         elif fuzzy_not(is_empty) and other.is_empty:
             return False
-        if self.is_finite is False and other.is_finite:
+        if self.is_finiteset is False and other.is_finiteset:
             return False
 
         # Dispatch on subclass rules
@@ -815,9 +815,9 @@ class ProductSet(Set):
         return fuzzy_or(s.is_empty for s in self.sets)
 
     @property
-    def is_finite(self):
-        all_finite = fuzzy_and(s.is_finite for s in self.sets)
-        return fuzzy_or(self.is_empty, all_finite)
+    def is_finiteset(self):
+        all_finite = fuzzy_and(s.is_finiteset for s in self.sets)
+        return fuzzy_or([self.is_empty, all_finite])
 
     @property
     def _measure(self):
@@ -1015,6 +1015,10 @@ class Interval(Set, EvalfMixin):
             cond = self.start > self.end  # Both bounds closed
         return fuzzy_bool(cond)
 
+    @property
+    def is_finiteset(self):
+        return self.measure.is_zero
+
     def _complement(self, other):
         if other == S.Reals:
             a = Interval(S.NegativeInfinity, self.start,
@@ -1188,6 +1192,10 @@ class Union(Set, LatticeOp, EvalfMixin):
     @property
     def is_empty(self):
         return fuzzy_and(set.is_empty for set in self.args)
+
+    @property
+    def is_finiteset(self):
+        return fuzzy_and(set.is_finiteset for set in self.args)
 
     @property
     def _measure(self):
@@ -1600,6 +1608,7 @@ class EmptySet(with_metaclass(Singleton, Set)):
     .. [1] https://en.wikipedia.org/wiki/Empty_set
     """
     is_empty = True
+    is_finiteset = True
     is_FiniteSet = True
 
     @property
@@ -1666,6 +1675,7 @@ class UniversalSet(with_metaclass(Singleton, Set)):
 
     is_UniversalSet = True
     is_empty = False
+    is_finiteset = False
 
     def _complement(self, other):
         return S.EmptySet
@@ -1718,7 +1728,7 @@ class FiniteSet(Set, EvalfMixin):
     is_FiniteSet = True
     is_iterable = True
     is_empty = False
-    is_finite = True
+    is_finiteset = True
 
     def __new__(cls, *args, **kwargs):
         evaluate = kwargs.get('evaluate', global_evaluate[0])

--- a/sympy/sets/sets.py
+++ b/sympy/sets/sets.py
@@ -12,7 +12,7 @@ from sympy.core.decorators import deprecated
 from sympy.core.evalf import EvalfMixin
 from sympy.core.evaluate import global_evaluate
 from sympy.core.expr import Expr
-from sympy.core.logic import fuzzy_bool, fuzzy_or, fuzzy_and
+from sympy.core.logic import fuzzy_bool, fuzzy_or, fuzzy_and, fuzzy_not
 from sympy.core.numbers import Float
 from sympy.core.operations import LatticeOp
 from sympy.core.relational import Eq, Ne
@@ -367,15 +367,17 @@ class Set(Basic):
             raise ValueError("Unknown argument '%s'" % other)
 
         # Handle the trivial cases
-        if self == other or self.is_empty:
+        if self == other:
             return True
+        is_empty = self.is_empty
+        if is_empty is True:
+            return True
+        elif fuzzy_not(is_empty) and other.is_empty:
+            return False
 
-        # Call the subclass method
-        ret = self._eval_is_subset(other)
-        if ret is not None:
-            return ret
-
-        ret = other._eval_is_superset(self)
+        # Use pairwise rules from multiple dispatch
+        from sympy.sets.handlers.issubset import is_subset_sets
+        ret = is_subset_sets(self, other)
         if ret is not None:
             return ret
 
@@ -385,18 +387,6 @@ class Set(Basic):
         # so that the intersect method uses is_subset for evaluation.
         if self.intersect(other) == self:
             return True
-
-    def _eval_is_subset(self, other):
-        """
-        Fuzzy bool indicating whether self is a subset of other
-        """
-        return None
-
-    def _eval_is_superset(self, other):
-        """
-        Fuzzy bool indicating whether self is a superset of other
-        """
-        return None
 
     # This should be deprecated:
     def issubset(self, other):
@@ -1057,46 +1047,6 @@ class Interval(Set, EvalfMixin):
         d = Dummy()
         return self.as_relational(d).subs(d, other)
 
-    def _eval_is_subset(self, other):
-        # Note that self.is_empty was already checked in Set.is_subset and can
-        # be assumed False now.
-
-        if isinstance(other, Interval):
-            # This is correct but can be made more comprehensive...
-            if fuzzy_bool(self.start < other.start):
-                return False
-            if fuzzy_bool(self.end > other.end):
-                return False
-            if (other.left_open and not self.left_open
-                    and fuzzy_bool(Eq(self.start, other.start))):
-                return False
-            if (other.right_open and not self.right_open
-                    and fuzzy_bool(Eq(self.end, other.end))):
-                return False
-
-        elif isinstance(other, FiniteSet):
-            # An Interval can only be a subset of a finite set if it is finite
-            # which can only happen if the start and end are equal (or if it
-            # is empty but that has already been checked).
-            if fuzzy_bool(self.start < self.end):
-                return False
-
-        elif isinstance(other, Union):
-            if all(isinstance(s, (Interval, FiniteSet)) for s in other.args):
-                intervals = [s for s in other.args if isinstance(s, Interval)]
-                if all(fuzzy_bool(self.start < s.start) for s in intervals):
-                    return False
-                if all(fuzzy_bool(self.end > s.end) for s in intervals):
-                    return False
-                if self.measure.is_nonzero:
-                    no_overlap = lambda s1, s2: fuzzy_or([
-                            fuzzy_bool(s1.end <= s2.start),
-                            fuzzy_bool(s1.start >= s2.end),
-                            ])
-                    if all(no_overlap(s, self) for s in intervals):
-                        return False
-
-
     def as_relational(self, x):
         """Rewrite an interval in terms of inequalities and logic operators."""
         x = sympify(x)
@@ -1293,13 +1243,6 @@ class Union(Set, LatticeOp, EvalfMixin):
 
     def _contains(self, other):
         return Or(*[s.contains(other) for s in self.args])
-
-    def _eval_is_subset(self, other):
-        return fuzzy_and(s.is_subset(other) for s in self.args)
-
-    def _eval_is_superset(self, other):
-        if fuzzy_or(s.is_superset(other) for s in self.args):
-            return True
 
     def as_relational(self, symbol):
         """Rewrite a Union in terms of equalities and logic operators. """
@@ -1666,12 +1609,6 @@ class EmptySet(with_metaclass(Singleton, Set)):
     def _contains(self, other):
         return false
 
-    def _eval_is_subset(self, other):
-        return True
-
-    def _eval_is_superset(self, other):
-        return other.is_empty
-
     def as_relational(self, symbol):
         return false
 
@@ -1871,14 +1808,6 @@ class FiniteSet(Set, EvalfMixin):
         # evaluate=True is needed to override evaluate=False context;
         # we need Eq to do the evaluation
         return fuzzy_or(fuzzy_bool(Eq(e, other, evaluate=True)) for e in self.args)
-
-    def _eval_is_subset(self, other):
-        return fuzzy_and(other._contains(e) for e in self.args)
-
-    def _eval_is_superset(self, other):
-        # A check for infiniteness would better...
-        if other in [S.Naturals, S.Naturals0, S.Integers, S.Rationals, S.Reals, S.Complexes]:
-            return False
 
     @property
     def _boundary(self):

--- a/sympy/sets/tests/test_fancysets.py
+++ b/sympy/sets/tests/test_fancysets.py
@@ -477,6 +477,7 @@ def test_Reals():
     assert S.Reals == Interval(-oo, oo)
     assert S.Reals != Interval(0, oo)
     assert S.Reals.is_subset(Interval(-oo, oo))
+    assert S.Reals.intersect(Range(-oo, oo)) == Range(-oo, oo)
 
 
 def test_Complex():

--- a/sympy/sets/tests/test_sets.py
+++ b/sympy/sets/tests/test_sets.py
@@ -616,6 +616,9 @@ def test_is_subset():
     assert FiniteSet(x).is_subset(FiniteSet(y).subs(y, x)) is True
     assert FiniteSet(x).is_subset(FiniteSet(y).subs(y, x+1)) is False
 
+    assert Interval(0, 1).is_subset(Interval(0, 1, left_open=True)) is False
+    assert Interval(-2, 3).is_subset(Union(Interval(-oo, -2), Interval(3, oo))) is False
+
 
 def test_is_proper_subset():
     assert Interval(0, 1).is_proper_subset(Interval(0, 2)) is True

--- a/sympy/sets/tests/test_sets.py
+++ b/sympy/sets/tests/test_sets.py
@@ -55,7 +55,34 @@ def test_imageset():
 def test_is_empty():
     for s in [S.Naturals, S.Naturals0, S.Integers, S.Rationals, S.Reals,
             S.UniversalSet]:
-        assert s.is_empty == False
+        assert s.is_empty is False
+
+    assert S.EmptySet.is_empty is True
+
+
+def test_is_finiteset():
+    for s in [S.Naturals, S.Naturals0, S.Integers, S.Rationals, S.Reals,
+            S.UniversalSet]:
+        assert s.is_finiteset is False
+
+    assert S.EmptySet.is_finiteset is True
+
+    assert FiniteSet(1, 2).is_finiteset is True
+    assert Interval(1, 2).is_finiteset is False
+    assert Interval(x, y).is_finiteset is None
+    assert ProductSet(FiniteSet(1), FiniteSet(2)).is_finiteset is True
+    assert ProductSet(FiniteSet(1), Interval(1, 2)).is_finiteset is False
+    assert ProductSet(FiniteSet(1), Interval(x, y)).is_finiteset is None
+    assert Union(Interval(0, 1), Interval(2, 3)).is_finiteset is False
+    assert Union(FiniteSet(1), Interval(2, 3)).is_finiteset is False
+    assert Union(FiniteSet(1), FiniteSet(2)).is_finiteset is True
+    assert Union(FiniteSet(1), Interval(x, y)).is_finiteset is None
+    assert Intersection(Interval(x, y), FiniteSet(1)).is_finiteset is True
+    assert Intersection(Interval(x, y), Interval(1, 2)).is_finiteset is None
+    assert Intersection(FiniteSet(x), FiniteSet(y)).is_finiteset is True
+    assert Complement(FiniteSet(1), Interval(x, y)).is_finiteset is True
+    assert Complement(Interval(x, y), FiniteSet(1)).is_finiteset is None
+    assert Complement(Interval(1, 2), FiniteSet(x)).is_finiteset is False
 
 
 def test_deprecated_is_EmptySet():

--- a/sympy/sets/tests/test_sets.py
+++ b/sympy/sets/tests/test_sets.py
@@ -63,26 +63,26 @@ def test_is_empty():
 def test_is_finiteset():
     for s in [S.Naturals, S.Naturals0, S.Integers, S.Rationals, S.Reals,
             S.UniversalSet]:
-        assert s.is_finiteset is False
+        assert s.is_finite_set is False
 
-    assert S.EmptySet.is_finiteset is True
+    assert S.EmptySet.is_finite_set is True
 
-    assert FiniteSet(1, 2).is_finiteset is True
-    assert Interval(1, 2).is_finiteset is False
-    assert Interval(x, y).is_finiteset is None
-    assert ProductSet(FiniteSet(1), FiniteSet(2)).is_finiteset is True
-    assert ProductSet(FiniteSet(1), Interval(1, 2)).is_finiteset is False
-    assert ProductSet(FiniteSet(1), Interval(x, y)).is_finiteset is None
-    assert Union(Interval(0, 1), Interval(2, 3)).is_finiteset is False
-    assert Union(FiniteSet(1), Interval(2, 3)).is_finiteset is False
-    assert Union(FiniteSet(1), FiniteSet(2)).is_finiteset is True
-    assert Union(FiniteSet(1), Interval(x, y)).is_finiteset is None
-    assert Intersection(Interval(x, y), FiniteSet(1)).is_finiteset is True
-    assert Intersection(Interval(x, y), Interval(1, 2)).is_finiteset is None
-    assert Intersection(FiniteSet(x), FiniteSet(y)).is_finiteset is True
-    assert Complement(FiniteSet(1), Interval(x, y)).is_finiteset is True
-    assert Complement(Interval(x, y), FiniteSet(1)).is_finiteset is None
-    assert Complement(Interval(1, 2), FiniteSet(x)).is_finiteset is False
+    assert FiniteSet(1, 2).is_finite_set is True
+    assert Interval(1, 2).is_finite_set is False
+    assert Interval(x, y).is_finite_set is None
+    assert ProductSet(FiniteSet(1), FiniteSet(2)).is_finite_set is True
+    assert ProductSet(FiniteSet(1), Interval(1, 2)).is_finite_set is False
+    assert ProductSet(FiniteSet(1), Interval(x, y)).is_finite_set is None
+    assert Union(Interval(0, 1), Interval(2, 3)).is_finite_set is False
+    assert Union(FiniteSet(1), Interval(2, 3)).is_finite_set is False
+    assert Union(FiniteSet(1), FiniteSet(2)).is_finite_set is True
+    assert Union(FiniteSet(1), Interval(x, y)).is_finite_set is None
+    assert Intersection(Interval(x, y), FiniteSet(1)).is_finite_set is True
+    assert Intersection(Interval(x, y), Interval(1, 2)).is_finite_set is None
+    assert Intersection(FiniteSet(x), FiniteSet(y)).is_finite_set is True
+    assert Complement(FiniteSet(1), Interval(x, y)).is_finite_set is True
+    assert Complement(Interval(x, y), FiniteSet(1)).is_finite_set is None
+    assert Complement(Interval(1, 2), FiniteSet(x)).is_finite_set is False
 
 
 def test_deprecated_is_EmptySet():

